### PR TITLE
fix: pull queries should work across nodes

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

Fixes: https://github.com/confluentinc/ksql/issues/4142
Fixes: https://github.com/confluentinc/ksql/issues/4151
Fixes: https://github.com/confluentinc/ksql/issues/4152

Introduces a new `inter.node.listener` that can be used to specify a URL that the node can be contacted on by other nodes.  This can be different to the listeners defined in `listeners`. This can be required if `listeners` is set to a wildcard address, i.e. IPv4 `0.0.0.0` or IPv6 `[::]`, or if the node sits behind network infrastructure that requires other nodes to reach it using a different URL.

If `inter.node.listener` is not set it still defaults to the first listener in `listener` config. However, it now replaces an wildcard address with `localHost`. This means inter-node comms is still possible for nodes running on the same host.

Warnings are logged if the inter-node listener resolves to a loopback or local address.

### Testing done 

Usual.  

Manual testing using two node docker KSQL cluster, using `KSQL_LISTENERS: http://0.0.0.0:8088`: the wildcard address is correctly replaced with `localhost` when used to share with other nodes. Optionally, can set `KSQL_ADVERTISED_LISTENER` to expose KSQL beyond local machine.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: config/ksql-production-server.properties [hunk 1]
```diff
@@ -18,17 +18,23 @@
 ### HTTP  ###
 # The URL the KSQL server will listen on:
 # The default is any IPv4 interface on the machine.
+# NOTE: If set to wildcard or loopback set 'inter.node.listener' to enable pull queries across machines
```
Comment (AlanConfluent) [inline config/ksql-production-server.properties]: https://kafka.apache.org/documentation/#configuration

We might want to be consistent with Kafka.  It seems like they use advertised.listeners.  Can we use a similar naming scheme?

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestConfig.java [hunk 2]
```diff
@@ -15,24 +15,50 @@
...
+  public static final String INTER_NODE_LISTENER_CONFIG =
+      KSQL_CONFIG_PREFIX + "inter.node.listener";
+  private static final String INTER_NODE_LISTENER_DOC =
+      "The listener used for communication between KSQL nodes in the cluster, if different to the '"
+          + LISTENERS_CONFIG + "' config property. "
+          + "In IaaS environments, this may need to be different from the interface to which "
+          + "the server binds. If this is not set, the first value from listeners will be used. "
+          + "Unlike listeners, it is not valid to use the 0.0.0.0 (IPv4) or [::] (IPv6) "
+          + "wildcard addresses.";
+  private static final String USE_FIRST_LISTENER = "http://use.first.listener";
```
Comment (AlanConfluent) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestConfig.java]: Do you need to have a default value that you don't intend to use as a default?  Seems like you could check for it being null or unset if that's the intention.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestConfig.java [hunk 3]
```diff
@@ -183,4 +220,170 @@ public KsqlRestConfig(final Map<?, ?> props) {
...
+                + "between nodes running on the same machine.",
+            sourceConfigName, listener
+        );
+      }
+    });
+
+    logger.info("Using {} config for intra-node communication: {}", sourceConfigName, listener);
+  }
+
+  private static URL sanitizeInterNodeListener(
```
Comment (AlanConfluent) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestConfig.java]: Add some comments indicating that this is only used for the "first listener" which may use the anyLocalAddress symbolic address.

**Comment by vpapavas** [comment]
Hey @big-andy-coates , should the `inter.node.listener` be the same as the streams config `application.server`?

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestConfig.java (hunk 2 above)
Comment (big-andy-coates) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestConfig.java]: Let me have another play. I actually intended to fix this before submitting, but forgot.

**Comment by big-andy-coates** [comment]
> Hey @big-andy-coates , should the `inter.node.listener` be the same as the streams config `application.server`?

It's _potentially_ used to _set_ `application.server`

**Comment by big-andy-coates** [comment]
Backported to 5.4.x in https://github.com/confluentinc/ksql/pull/4271.  (Should be released in 5.4.1)

**Pull request merged**